File: src/java/com/distributedlife/mahjong/reference/permute/MultiplePermutator.java
package com.distributedlife.mahjong.reference.permute;

import com.distributedlife.mahjong.reference.data.TileSet;
import com.distributedlife.mahjong.reference.hand.HandCandidate;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.distributedlife.mahjong.reference.hand.HandCandidate.times;

public class MultiplePermutator extends Permutator {
    final int multiples;
    private final String defaultSuit;
    final List<String> tilesMultipleIsAllowedIn;

    public MultiplePermutator(List<String> tilesMultipleIsAllowedIn, int multiples, String defaultSuit) {
        this.tilesMultipleIsAllowedIn = tilesMultipleIsAllowedIn;
        this.multiples = multiples;
        this.defaultSuit = defaultSuit;
    }

    @Override
    public Set<HandCandidate> permute(Set<HandCandidate> candidates) {
        return permute(candidates, defaultSuit);
    }

    @Override
    public Set<HandCandidate> permute(Set<HandCandidate> candidates, String suit) {
        Set<HandCandidate> handCandidates = new HashSet<HandCandidate>();

        for(HandCandidate candidate : candidates) {
            for(String tile : tilesMultipleIsAllowedIn) {
                String tileInSuit = TileSet.convertTileToAppropriateSuit(suit, tile, candidate);

                if (candidate.canAdd(tileInSuit, times(multiples))) {
                    HandCandidate variation = candidate.fork();
                    variation.add(tileInSuit, times(multiples));
                    handCandidates.add(variation);
                }
            }
        }

        return handCandidates;
    }
}

File: lib/permute/multiple-permutator.js
/**
 * MultiplePermutator - Handles permutations for pairs, pungs, and kongs
 */
import { Permutator } from './permutator.js';
import { convertTileToAppropriateSuit } from '../data/tile-set.js';
import { times } from '../hand/hand-candidate.js';

export class MultiplePermutator extends Permutator {
  constructor(tilesMultipleIsAllowedIn, multiples, defaultSuit) {
    super();
    this.tilesMultipleIsAllowedIn = tilesMultipleIsAllowedIn;
    this.multiples = multiples;
    this.defaultSuit = defaultSuit;
  }

  permute(candidates, suit = null) {
    const useSuit = suit || this.defaultSuit;
    const handCandidates = new Set();

    for (const candidate of candidates) {
      for (const tile of this.tilesMultipleIsAllowedIn) {
        const tileInSuit = convertTileToAppropriateSuit(useSuit, tile, candidate);

        if (candidate.canAdd(tileInSuit, times(this.multiples))) {
          const variation = candidate.fork();
          variation.add(tileInSuit, times(this.multiples));
          handCandidates.add(variation);
        }
      }
    }

    return handCandidates;
  }
}
